Deferred/KSPromise.h

```diff
@@ -2,27 +2,25 @@
 #import "KSCancellable.h"
 #import "KSNullabilityCompat.h"
 
-@class KSPromise;
-
 NS_ASSUME_NONNULL_BEGIN
 
-typedef __nonnull id(^promiseValueCallback)(id value);
-typedef __nonnull id(^promiseErrorCallback)(NSError *error);
-typedef void(^deferredCallback)(KSPromise *p);
-
 FOUNDATION_EXPORT NSString *const KSPromiseWhenErrorDomain;
 FOUNDATION_EXPORT NSString *const KSPromiseWhenErrorErrorsKey;
 FOUNDATION_EXPORT NSString *const KSPromiseWhenErrorValuesKey;
 
-@interface KSPromise : NSObject<KSCancellable>
-@property (strong, nonatomic, readonly, nullable) id value;
+@interface KSPromise<ObjectType> : NSObject<KSCancellable>
+typedef __nonnull id(^promiseValueCallback)(ObjectType value);
+typedef __nonnull id(^promiseErrorCallback)(NSError *error);
+typedef void(^deferredCallback)(KSPromise<ObjectType> *p);
+
+@property (strong, nonatomic, readonly, nullable) ObjectType value;
 @property (strong, nonatomic, readonly, nullable) NSError *error;
 @property (assign, nonatomic, readonly) BOOL fulfilled;
 @property (assign, nonatomic, readonly) BOOL rejected;
 @property (assign, nonatomic, readonly) BOOL cancelled;
 
 + (KSPromise *)when:(NSArray *)promises;
-- (KSPromise *)then:(nullable promiseValueCallback)fulfilledCallback error:(nullable promiseErrorCallback)errorCallback;
+- (KSPromise *)then:(nullable id(^)(ObjectType value))fulfilledCallback error:(nullable promiseErrorCallback)errorCallback;
```

Comment:
I don't anticipate seeing any more extensions to this functionality in Xcode 7, or even later Xcode versions, to be honest. I'm reasonably certain that the Apple's primary motivation in introducing this feature was to improve the quality of the Swift bridging, and what they are currently providing is sufficient for the needs of Apple's frameworks, as far as I can tell.

It's also interesting to consider how all the stronger typing that Apple has been introducing to Obj-C over the past couple years is just taking it further away from its Smalltalk roots. If you're looking for real strong typing and editor assistance for your monadic binds...well, Swift is there waiting for you to jump on board.